src/main/java/minijava/ir/assembler/block/CodeBlock.java

```diff
@@ -0,0 +1,224 @@
+package minijava.ir.assembler.block;
+
+import static com.google.common.base.Preconditions.checkArgument;
+
+import com.google.common.collect.Iterators;
+import java.util.*;
+import java.util.stream.Collectors;
+import minijava.ir.assembler.GNUAssemblerConvertible;
+import minijava.ir.assembler.instructions.Cmp;
+import minijava.ir.assembler.instructions.Instruction;
+import minijava.ir.assembler.instructions.Jmp;
+import org.jetbrains.annotations.NotNull;
+
+/**
+ * A list of assembler instructions with a label, optional cmp support, cmp and jmp instructions.
+ */
+public class CodeBlock implements GNUAssemblerConvertible, Collection<Instruction> {
```

Comment:
Again: Implementing the whole `Collection<>` is more misleading than anything, especially if all you ever do is `add` or check `isEmpty`. Especially since you throw `UnsupportedOperationException` most of the time.

If you *really* want to have the collection interface, expose the `normalInstructions` field in a getter.

Generally implementing `Collection`/`List`/`Iterable` etc. interfaces is a code smell. Most of the time you can the relevant logic into a collection class with a single responsibility (as opposed to these frankensteins) or even just expose a field.